src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from typing import ClassVar, Generator, Pattern
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.option.global_options import GlobalOptions
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+_pex_target_regexp: str = r"""
+(?# optional path, one level with dot-separated words)
+(?:(?P<path>(?:\w[.0-9_-]?)+) /)?
+
+(?# binary name, with .pex file extension)
+(?P<name>(?:\w[.0-9_-]?)+) \.pex$
+"""
+
+
+@dataclass(frozen=True)
+class ParsedDockerfile:
+    commands: tuple[Command, ...]
+    pex_target_regexp: ClassVar[Pattern]
+
+    def __post_init__(self):
+        # Compile regexp when creating the first instance, and store it on the class.
+        if not getattr(ParsedDockerfile, "pex_target_regexp", None):
+            ParsedDockerfile.pex_target_regexp = re.compile(_pex_target_regexp, re.VERBOSE)
+
+    @classmethod
+    def parse(cls, dockerfile: str) -> "ParsedDockerfile":
```

Comment:
Maybe `dockerfile_contents`? So it's obvious that you pass the contents, not a path?